From 17a794fd0b4af762ce6c782a8e5473908c127dc0 Mon Sep 17 00:00:00 2001
From: Shivani Raval <sraval@habana.ai>
Date: Fri, 04 Apr 2025 10:15:47 +0300
Subject: [PATCH] [JSW-31] preboot: added files for jgs

As of now added the same code as FS1 to pass
JGS asic build.

Change-Id: I64ca22dc9512b48d3db65be4df2efd07288608d2
---

diff --git a/arch/jgs/arc/asic.c b/arch/jgs/arc/asic.c
new file mode 100644
index 0000000..84c3a9b
--- /dev/null
+++ b/arch/jgs/arc/asic.c
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier:     Apache-2.0
+/*
+ * (C) Copyright 2025 Intel, Ltd.
+ */
+
+#include <address_space.h>
+#include <preboot.h>
+#include <pll.h>
+#include <hl.h>
+#include <cmd_module.h>
+
+/* timestamp fed with constand ref_clk/2 */
+#define TS_CLOCK_FREQUENCY	25000000ULL
+
+u64 hl_pll_timestamp_freq_get(void)
+{
+	return TS_CLOCK_FREQUENCY;
+}
+
+void hl_dev_ts_init(void)
+{
+	/* TODO: implement ts init for JGS */
+}
+
+void hl_ts_shutdown(void)
+{
+	/* TODO: implement ts shutdown for JGS */
+}
+
+void pci_dev_specific_hot_rst(void)
+{
+	/* TODO: implement hot rst for JGS */
+}
+
+bool is_pcie_initialized(void)
+{
+	return true;
+}
+
+void preboot_plat_early_init(void) {}
+
+void preboot_security_init(void)
+{
+	/* TODO: implement security init for JGS */
+}
diff --git a/arch/jgs/arc/common.c b/arch/jgs/arc/common.c
new file mode 100644
index 0000000..dadbb09
--- /dev/null
+++ b/arch/jgs/arc/common.c
@@ -0,0 +1,501 @@
+// SPDX-License-Identifier:     Apache-2.0
+/*
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#include <psoc_boot_conf_regs.h>
+#include <psoc_boot_conf_masks.h>
+#include <arc_regs.h>
+#include <ca53_cfg_regs.h>
+#include <global_conf_regs.h>
+#include <i2c_regs.h>
+#include <q_h2d_regs.h>
+#include <q_d2h_regs.h>
+#include <muc_arc_aux_regs.h>
+#include <nocp_rtr_cfg_regs.h>
+#include <jgs_blocks_linux_driver.h>
+#include <gpu_utils.h>
+
+#include <preboot.h>
+#include <types.h>
+#include <common.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <boot_if.h>
+#include <hl_device.h>
+#include <jgs_reg_map.h>
+#include <hl_platform_defs.h>
+#include <hash_defs.h>
+#include <sha512.h>
+#include <hl_fw_upd_if.h>
+#include <hl_pcie.h>
+#include <jgs_fw_if.h>
+#include <hl_fit_if.h>
+#include "hl.h"
+#include <hl_boot_desc.h>
+#include <boot_desc.h>
+
+#include <arc_local_conf_regs.h>
+#include <arc_local_conf_masks.h>
+#include <hl_devs.h>
+#include <pcie_dbi_regs.h>
+#include <pcie_dbi_masks.h>
+#include <pcie_aux_regs.h>
+#include <pcie_core_regs.h>
+#include <bmu_regs.h>
+#include <bmu_masks.h>
+#include <h15b_psoc_reset_conf_regs.h>
+#include <pcie_flr_regs.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(fs1_arc_common);
+
+static void *masks __aligned(4);
+
+void halt_n_reset_arc(uint8_t id)
+{
+	/* TODO: add implementation */
+}
+
+u32 hl_rst_src_get(void)
+{
+	return CFG_RREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_RST_SRC));
+}
+
+uint32_t preboot_get_nic_mask(void)
+{
+	/* TODO: add nic mask for JGS */
+	return 0;
+}
+
+void preboot_nic_en_hw(u32 nic_mask)
+{
+	/* TODO: implement for JGS */
+}
+
+void device_init_pci(void)
+{
+	pci.dbi_base_addr = mmIO_0_CFG_CXL_EP_DBI_SIG_BASE;
+	pci.dbi_bar0_reg = mmIO_0_CFG_CXL_EP_DBI_SIG_BASE + mmPCIE_DBI_BAR0_REG;
+	pci.dbi_bar2_reg = mmIO_0_CFG_CXL_EP_DBI_SIG_BASE + mmPCIE_DBI_BAR2_REG;
+	pci.dbi_bar4_reg = mmIO_0_CFG_CXL_EP_DBI_SIG_BASE + mmPCIE_DBI_BAR4_REG;
+	pci.dbi_status_command_reg = mmIO_0_CFG_CXL_EP_DBI_SIG_BASE +
+				     mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.aux_dbi_reg = mmIO_0_CFG_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_DBI;
+	pci.aux_mem_space_en_reg = mmIO_0_CFG_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_MEM_SPACE_EN;
+	pci.aux_flr_ctrl_reg = mmIO_0_CFG_CXL_EP_PCIE_FLR_BASE + mmPCIE_FLR_FLR_CONTROL;
+	pci.aux_ltssm_en_reg = mmIO_0_CFG_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_LTSSM_EN;
+	pci.aux_pf_flr_active_reg = mmIO_0_CFG_CXL_EP_PCIE_FLR_BASE +
+				    mmPCIE_FLR_CFG_FLR_PF_ACTIVE;
+	pci.aux_flr_int_reg = mmIO_0_CFG_CXL_EP_PCIE_FLR_BASE +
+			      mmPCIE_FLR_INTR_SPI_CAUSE_CFG_FLR_PF_ACTIVE;
+	pci.aux_pf_flr_done_reg = mmIO_0_CFG_CXL_EP_PCIE_FLR_BASE +
+				  mmPCIE_FLR_APP_FLR_PF_DONE;
+	pci.aux_rdlh_link_up_reg = mmIO_0_CFG_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_RDLH_LINK_UP;
+	pci.aux_smlh_link_up_reg = mmIO_0_CFG_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_SMLH_LINK_UP;
+	pci.core_hot_rst_reg = mmIO_0_CFG_CXL_EP_PCIE_CORE_BASE +
+			       mmPCIE_CORE_INTR_SPI_CAUSE_HOT_RESET;
+	pci.bmu_base_addr = mmIO_0_CFG_CXL_EP_PCIE_BMU_CFG_BASE;
+}
+
+void device_init_global_conf(void)
+{
+	global_conf.prstn_reg = IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_PRSTN);
+	global_conf.i2c_slv_reg = IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_I2C_SLV);
+	global_conf.i2c_slv_intr_mask_reg = mmIO_0_CFG_PSOC_PSOC_CFG_I2C_S_BASE +
+					    mmI2C_IC_INTR_MASK;
+	global_conf.cpu_boot_status_reg = IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_STATUS);
+	global_conf.upd_pending_sts_map = mmUPD_PENDING_STS;
+	global_conf.upd_cmd_map = mmUPD_CMD;
+	global_conf.upd_sts_map = mmUPD_STS;
+	global_conf.hw_state_reg = mmHW_STATE;
+}
+
+void device_init_ca53_cfg(void)
+{
+	ca53_cfg.arm_rst_control_reg = mmIO_0_CFG_SOC_NOC_CPU_CA53_CFG_BASE +
+				       mmCA53_CFG_ARM_RST_CONTROL;
+}
+
+void lkd_dev_specific_setup(void)
+{
+	cpu_boot_dev_sts_set(CPU_BOOT_DEV_STS_FW_HARD_RST_EN);
+}
+
+void lkd_dev_spec_regs(struct cpu_dyn_regs *regs)
+{
+	regs->pq_block_base_addr = mmIO_0_CFG_PSOC_ARC_MNG_AUX_Q0_H2D_BASE;
+	regs->cq_block_base_addr = mmIO_0_CFG_PSOC_ARC_MNG_AUX_Q0_D2H_BASE;
+	regs->eq_block_base_addr = mmIO_0_CFG_PSOC_ARC_MNG_AUX_Q1_D2H_BASE;
+	regs->cpu_boot_dev_sts0 = mmCPU_BOOT_DEV_STS0;
+	regs->cpu_boot_dev_sts1 = mmCPU_BOOT_DEV_STS1;
+	regs->cpu_boot_err0 = mmCPU_BOOT_ERR0;
+	regs->cpu_boot_err1 = mmCPU_BOOT_ERR1;
+	regs->cpu_boot_status = IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_STATUS);
+	regs->fw_upd_sts = mmUPD_STS;
+	regs->fw_upd_cmd = mmUPD_CMD;
+	regs->fw_upd_pending_sts = mmUPD_PENDING_STS;
+	regs->hw_state = mmHW_STATE;
+	regs->kmd_msg_to_cpu = IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_KMD_MSG_TO_CPU);
+	regs->cpu_cmd_status_to_host = mmCPU_CMD_STATUS_TO_HOST;
+}
+
+void cpu_boot_if_dev_spec_regs_init(struct cpu_boot_if_regs *regs)
+{
+	/* boot_if module wants the full address */
+	regs->status = HL_CFG_BASE + IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_STATUS);
+	regs->response = HL_CFG_BASE + mmCPU_CMD_STATUS_TO_HOST;
+	regs->kmd_msg_to_cpu = HL_CFG_BASE + IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_KMD_MSG_TO_CPU);
+	regs->hw_state = HL_CFG_BASE + mmHW_STATE;
+	regs->err[0] = HL_CFG_BASE + mmCPU_BOOT_ERR0;
+	regs->err[1] = HL_CFG_BASE + mmCPU_BOOT_ERR1;
+	regs->dev_sts[0] = HL_CFG_BASE + mmCPU_BOOT_DEV_STS0;
+	regs->dev_sts[1] = HL_CFG_BASE + mmCPU_BOOT_DEV_STS1;
+}
+
+bool hl_is_debug_mode(enum hl_debug_hw_func func)
+{
+	u32 boot_strap = regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_L));
+	u32 strap_mask;
+
+	switch (func) {
+	case HL_DEBUG_HW_PCIE:
+		strap_mask = PSOC_BOOT_CONF_BOOT_STRAP_PINS_L_BTL_DISABLE_M; /* BTL_DISABLE */
+		break;
+	case HL_DEBUG_HW_DRAM:
+		strap_mask = BIT(10); /* RSVD[1] */
+		break;
+	case HL_DEBUG_HW_I2C:
+		strap_mask = BIT(9); /* RSVD[0] */
+		break;
+	default:
+		LOG_ERR("Unknown H/W function %d", func);
+		return 0;
+	}
+
+	return boot_strap & strap_mask;
+}
+
+bool preboot_sram_is_valid_address_range(uint64_t addr, int len)
+{
+	if (addr >= IO0_SP_SRAM_BASE && addr + len < IO0_SP_SRAM_BASE + IO_SP_SRAM_SIZE)
+		return true;
+
+	return false;
+}
+
+bool preboot_is_valid_address_range(u64 addr, int len)
+{
+	if (addr >= IO0_SP_SRAM_BASE && addr + len < IO0_SP_SRAM_BASE + IO_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= IO0_SP_SRAM_FULL_BASE && addr + len <= IO0_SP_SRAM_FULL_BASE + IO_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= IO0_PID_ICCM_FULL_BASE &&
+	    addr + len <= IO0_PID_ICCM_FULL_BASE + IO_PID_ICCM_SIZE)
+		return true;
+
+	if (addr >= IO0_PID_DCCM_FULL_BASE &&
+	    addr + len <= IO0_PID_DCCM_FULL_BASE + IO_PID_DCCM_SIZE)
+		return true;
+
+	if (addr >= IO0_MNG_ICCM_FULL_BASE &&
+	    addr + len <= IO0_MNG_ICCM_FULL_BASE + IO_MNG_ICCM_SIZE)
+		return true;
+
+	if (addr >= IO0_MNG_DCCM_FULL_BASE &&
+	    addr + len <= IO0_MNG_DCCM_FULL_BASE + IO_MNG_DCCM_SIZE)
+		return true;
+
+	if (addr >= IO0_MISC_ICCM_FULL_BASE &&
+	    addr + len <= IO0_MISC_ICCM_FULL_BASE + IO_MISC_ICCM_SIZE)
+		return true;
+
+	if (addr >= IO0_MISC_DCCM_FULL_BASE &&
+	    addr + len <= IO0_MISC_DCCM_FULL_BASE + IO_MISC_DCCM_SIZE)
+		return true;
+
+	if (addr >= IO0_RST_ICCM_FULL_BASE &&
+	    addr + len <= IO0_RST_ICCM_FULL_BASE + IO_RST_ICCM_SIZE)
+		return true;
+
+	if (addr >= IO0_RST_DCCM_FULL_BASE &&
+	    addr + len <= IO0_RST_DCCM_FULL_BASE + IO_RST_DCCM_SIZE)
+		return true;
+
+	if (addr >= IO0_MUC_ICCM_FULL_BASE &&
+	    addr + len <= IO0_MUC_ICCM_FULL_BASE + IO_MUC_ICCM_SIZE)
+		return true;
+
+	if (addr >= IO0_MUC_DCCM_FULL_BASE &&
+	    addr + len <= IO0_MUC_DCCM_FULL_BASE + IO_MUC_DCCM_SIZE)
+		return true;
+
+	LOG_ERR("ERROR: invalid address range addr = 0x%016llx", addr);
+
+	return false;
+}
+
+void hl_set_prompt_name(char *prompt_name)
+{
+	snprintf(prompt_name, sizeof(HL_SHELL_PROMPT_NAME), HL_SHELL_PROMPT_NAME, '0');
+}
+
+bool preboot_is_core_id_supported(void)
+{
+	return false;
+}
+
+u32 hl_pll_timestamp_base_get(void)
+{
+	return mmIO_0_CFG_PSOC_PSOC_CFG_TIMESTAMP_BASE;
+}
+
+u32 lkd_dev_spec_get_desc_offset(uintptr_t base)
+{
+	if (base >= IO0_SP_SRAM_BASE && base < IO0_SP_SRAM_BASE + IO_SP_SRAM_SIZE)
+		return ((u32)base - IO0_SP_SRAM_BASE);
+
+	/* TODO: implement addr conversion for HBM (FSW-3487) */
+
+	LOG_ERR("%s: unsupported addr %p", __func__, (void *)base);
+
+	return 0;
+}
+
+void sha_csum(const unsigned char *input, unsigned int ilen, unsigned char *output)
+{
+	mbedtls_sha512(input, ilen, output, MBEDTLS_USE_SHA384);
+}
+
+void preboot_read_straps(void)
+{
+	u32 boot_strap_l = CFG_RREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_L));
+	u32 boot_strap_h = CFG_RREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_H));
+
+	LOG_INF("Boot strap pins low: 0x%08x, high: 0x%08x",
+		boot_strap_l, boot_strap_h);
+}
+
+
+static int preboot_isolation(enum fs1_die_id die_id)
+{
+	/* TODO: set isolation for jgs */
+
+	return 0;
+}
+
+int preboot_isolation_init(void)
+{
+	/* TODO: init isolation for jgs*/
+	return 0;
+}
+
+u32 preboot_get_elbi_reg_offset_from_base(uintptr_t address)
+{
+	return (u32)((address - IO0_SP_SRAM_BASE) & ~HL_UPD_STS_OFFSET_MASK);
+}
+
+u32 dev_get_fwupd_status_offset(uintptr_t address)
+{
+	u32 offset;
+
+	if (address >= IO0_SP_SRAM_BASE)
+		offset = (u32)((address - IO0_SP_SRAM_BASE) & ~HL_UPD_STS_OFFSET_MASK);
+	else
+		offset = (u32)(address & ~HL_UPD_STS_OFFSET_MASK);
+
+	return offset;
+}
+
+static uint64_t get_hbm_size(void)
+{
+	/* TODO: get HBM size depending on HBM type and binning */
+	return (FS1_NUM_OF_HBMS * FS1_HBM_MODULE_SIZE);
+}
+
+__weak int get_bmu_development_phase(enum bmu_phase *dev_phase)
+{
+	*dev_phase = BMU_PRODUCTION_MODE;
+
+	return 0;
+}
+
+static int get_bmu_configs(struct bmu_configs *conf)
+{
+	int err;
+	uint32_t val;
+
+	conf->hbm_size_mb = get_hbm_size();
+	val = CFG_RREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_NUM_VFS)) & PCIE_DBI_SRIOV_NUM_VFS_VAL_M;
+	switch (val) {
+	case 0:
+		conf->sriov = BMU_1PF_MODE;
+		break;
+	case 1:
+		conf->sriov = BMU_1PF_1VF_MODE;
+		break;
+	case 4:
+		conf->sriov = BMU_1PF_4VF_MODE;
+		break;
+	default:
+		LOG_ERR("unsupported sriov mode num_vf=%d", val);
+		return -1;
+	}
+
+	err = get_bmu_development_phase(&conf->phase);
+
+	return err;
+}
+
+void hl_pci_bmu_init(void)
+{
+	int err;
+	struct bmu_configs conf;
+
+	err = get_bmu_configs(&conf);
+	if (err) {
+		LOG_ERR("bmu configs not valid, skipping BMU init");
+		return;
+	}
+
+	err = fs1_bmu_init(&conf);
+
+	/* if configuration is valid, enable the BMUs */
+	if (!err) {
+		CFG_RMWREG32(IO0_H0_BMU_CFG(mmBMU_CFG_BMU), 1, BMU_CFG_BMU_EN_M);
+		CFG_RMWREG32(IO0_H0_BMU_MEM(mmBMU_CFG_BMU), 1, BMU_CFG_BMU_EN_M);
+	}
+}
+
+u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	if (base >= IO0_SP_SRAM_BASE && base < IO0_SP_SRAM_BASE + IO_SP_SRAM_SIZE)
+		return (base & ~IO0_SP_SRAM_BASE) + IO0_SP_SRAM_FULL_BASE;
+
+	/* TODO: implement addr conversion for HBM (FSW-3487) */
+
+	LOG_ERR("%s: unsupported addr %p", __func__, (void *)base);
+
+	return 0;
+}
+
+void preboot_copy_image(u64 dst, const void *buf, int len, const char *arch)
+{
+	u32 arc_dst = fs1_get_arc_addr_from_64bit(dst);
+
+	memcpy((void *)arc_dst, buf, len);
+}
+
+void image_start(u64 loadaddr)
+{
+	u32 entry = fs1_get_arc_addr_from_64bit(loadaddr);
+	void (*image_entry)(void) = (void *)entry;
+
+	image_entry();
+}
+
+void preboot_start_image(u64 startaddr, bool valid_arc_img, enum fit_core_id core_id)
+{
+	if (preboot_is_core_id_supported() && (core_id != FIT_CORE_ID_IO0_MGMT)) {
+		/* core-id mode is supported only with ARC.MNG of DIE-0 */
+		LOG_ERR("no start function for core_id = %d", core_id);
+		return;
+	}
+
+	image_start(startaddr);
+}
+
+static int jgs_engines_binning_init(struct fs1_binning_info *info)
+{
+	/* TODO: binning init for jgs*/
+	return 0;
+}
+
+__weak int preboot_plat_binning_init(void)
+{
+	/* TODO: binning init for jgs*/
+	return 0;
+}
+
+int preboot_binning_init(void)
+{
+	return preboot_plat_binning_init();
+}
+
+u32 _dev_spec_get_img_desc(struct hl_fw_img_desc *img_desc, enum fw_post_reset_cmd upd_status,
+			  enum img_idx sel_idx_mode, struct boot_desc *desc)
+{
+	/* TODO: use stable index when bootinfo implementation is final in arcboot */
+	desc->spi_offset = img_desc->spi_offset[0];
+	desc->img_size = img_desc->size;
+	desc->loadaddr = ((u64)img_desc->load_high << 32) | img_desc->load_low;
+	desc->desc_ver = img_desc->desc_ver;
+	desc->crc32 = img_desc->crc32;
+	memcpy(desc->chksum, img_desc->chksum, ARRAY_SIZE(desc->chksum));
+
+	return 0;
+}
+
+/*
+ * TODO: implement this function properly (halt, reset, release reset, scrub memories) using CPU
+ * parameters from fit file, as required in FSW-8421
+ */
+void preboot_configure_arc(int id, uint64_t loadaddr)
+{
+	switch (id) {
+	case FIT_CORE_ID_IO0_MUC:
+		CFG_WREG32(mmIO_0_CFG_MEDIA_0_MUC_AUX_BASE + mmMUC_ARC_AUX_ARC_RST, 0);
+		break;
+	default:
+		LOG_ERR("%s unsupported arc id %d", __func__, id);
+	}
+}
+
+static u32 io0_rtr_with_sa_switch[] = { /* TODO */
+					};
+
+#define PCI_HBW_ADEC_EN		(BIT(12))
+
+static void init_standalone_routing(void)
+{
+	u32 rtr_b;
+	int i;
+
+	LOG_INF("IO standalone mode datapath config");
+
+	/* config PSOC to STANDALONE communication mode */
+	CFG_WREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_STAND_ALONE_MODE), 1);
+	CFG_WREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_PCIE_STAND_ALONE_MODE), 1);
+
+	/* in standalone IO direct all HBW traffic to PCI */
+	for (i = 0; i < ARRAY_SIZE(io0_rtr_with_sa_switch); i++) {
+		rtr_b = io0_rtr_with_sa_switch[i];
+		CFG_WREG32(rtr_b + mmNOCP_RTR_CFG_SA_HBW_ADEC_MASK_L_12, 0);
+		CFG_WREG32(rtr_b + mmNOCP_RTR_CFG_SA_HBW_ADEC_MASK_H_12, 0);
+		CFG_WREG32(rtr_b + mmNOCP_RTR_CFG_SA_HBW_ADEC_BASE_L_12, 0);
+		CFG_WREG32(rtr_b + mmNOCP_RTR_CFG_SA_HBW_ADEC_BASE_H_12, 0);
+		CFG_WREG32(rtr_b + mmNOCP_RTR_CFG_SA_HBW_ADEC_EN, PCI_HBW_ADEC_EN);
+	}
+}
+
+void preboot_security_dynamic_init(void) {}
+void preboot_nic_mask_serr_derr_interrupts(void) {}
+void preboot_reset_init(void) {}
+void preboot_clear_pcie_mstr_axuser_cfg(void) {}
+void preboot_read_test_mode(void) {}
+void hl_set_reset_cause(uint32_t reset_cause) {}
+void preboot_dev_spec_bootinfo_init(void) {}
+
+int preboot_get_binning_info(void)
+{
+	return 0;
+}
+
+void preboot_engines_init(void)
+{
+	if (dev_is_standalone_io())
+		init_standalone_routing();
+}
diff --git a/arch/jgs/arc/common.h b/arch/jgs/arc/common.h
new file mode 100644
index 0000000..c965bd8
--- /dev/null
+++ b/arch/jgs/arc/common.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier:     Apache-2.0
+ *
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#ifndef COMMON_JGS_ARC_GNU_H_
+#define COMMON_JGS_ARC_GNU_H_
+
+#include <zephyr/sys/util.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/misc/hl_hbm/common/hl_hbm.h>
+#include <zephyr/drivers/misc/hl_hbm/jgs/jgs_hbm.h>
+#include <jgs_blocks_linux_driver.h>
+
+#define ADDR_MASK(_mem_type)		(_mem_type ## _SIZE - 1)
+
+#define HL_EROM_BOOT_PTRS_BASE		0
+#define HL_FW_BOOT_PTRS_BASE		0
+
+/* zephyr bin base address and size on spsram */
+#define PREBOOT_BIN_START              DT_REG_ADDR(DT_INST(0, mmio_sram))
+#define PREBOOT_BIN_SIZE               (DT_REG_SIZE(DT_INST(0, mmio_sram)) * SZ_1K)
+#define FUSE_WORDS_PER_BANK		0x80
+#define I2C_SLAVE_ADDRESS		0x6A
+
+/* TODO: replace with true TS calibration values when available */
+#define TS_UNCALIB_A_G_COEFF		0
+#define TS_UNCALIB_B_H_COEFF		0
+#define TS_CALIB_A_G_COEFF		0
+#define TS_CALIB_B_H_COEFF		0
+
+#define WATCHDOG_DEFAULT_TIMEOUT_MS	(120 * 1000)
+
+#define HL_SHELL_PROMPT_NAME	"OAM[%c] Fs1>"
+
+#endif /* COMMON_JGS_ARC_GNU_H_ */
diff --git a/arch/jgs/arc/hl.h b/arch/jgs/arc/hl.h
new file mode 100644
index 0000000..66a0458
--- /dev/null
+++ b/arch/jgs/arc/hl.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier:     Apache-2.0
+ *
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#ifndef ARCH_FS1_ARC_HL_H_
+#define ARCH_FS1_ARC_HL_H_
+
+#include <jgs_fw_if.h>
+#include <hl_platform_defs.h>
+
+void halt_n_reset_arc(u8 id);
+int jgs_redundancy_ctx_init(void *ctx, enum fs1_die_id die_id);
+
+#endif /* ARCH_FS1_COMMON_HL_H_ */
diff --git a/arch/jgs/arc/memory_dev.h b/arch/jgs/arc/memory_dev.h
new file mode 100644
index 0000000..44d1ecb
--- /dev/null
+++ b/arch/jgs/arc/memory_dev.h
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier:     Apache-2.0
+ *
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#ifndef ARCH_FS1_MEMORY_DEV_H_
+#define ARCH_FS1_MEMORY_DEV_H_
+
+extern int _image_ram_end; /* linker symbol, the end of the memory occupied by preboot */
+
+/*
+ * the following macro is taken from the FS1 linker file at zephyr/soc/arc/hl/fs1/linker.ld
+ * please handle changes with care and sync with the linker file.
+ */
+#if defined(CONFIG_PREBOOT_RUN_FROM_ICCM)
+	#if DT_NODE_HAS_PROP(DT_INST(0, arc_iccm), reg) && \
+		(DT_REG_SIZE(DT_INST(0, arc_iccm)) > 0)
+	#define SRAM_START		DT_REG_ADDR(DT_INST(0, arc_iccm))
+	#define SRAM_SIZE		(DT_REG_SIZE(DT_INST(0, arc_iccm)) * SZ_1K)
+	#endif
+#else
+	/* SRAM base address and size */
+	#if DT_NODE_HAS_PROP(DT_INST(0, mmio_sram), reg) && \
+		(DT_REG_SIZE(DT_INST(0, mmio_sram)) > 0)
+		#if DT_NODE_HAS_PROP(DT_NODELABEL(bootinfo_partition), reg) && \
+			(DT_REG_ADDR(DT_NODELABEL(bootinfo_partition)) == 0)
+			#define SRAM_START	(DT_REG_ADDR(DT_INST(0, mmio_sram)) + \
+				(DT_REG_SIZE(DT_NODELABEL(bootinfo_partition)) * SZ_1K))
+			#define SRAM_SIZE	((DT_REG_SIZE(DT_INST(0, mmio_sram)) - \
+				DT_REG_SIZE(DT_NODELABEL(bootinfo_partition))) * SZ_1K)
+		#else
+			#define SRAM_START	DT_REG_ADDR(DT_INST(0, mmio_sram))
+			#define SRAM_SIZE	(DT_REG_SIZE(DT_INST(0, mmio_sram)) * SZ_1K)
+		#endif
+	#endif
+#endif
+
+/* default heap is allocated right after preboot app */
+#define MEM_HEAP_BASE	((unsigned int)&_image_ram_end)
+#define MEM_HEAP_SZ	(SRAM_SIZE - ((unsigned int)&_image_ram_end - SRAM_START))
+
+#endif /* ARCH_GAUDI2_MEMORY_DEV_H_ */
diff --git a/arch/jgs/arc/onix.c b/arch/jgs/arc/onix.c
new file mode 100644
index 0000000..9e2675e
--- /dev/null
+++ b/arch/jgs/arc/onix.c
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier:     Apache-2.0
+/*
+ * (C) Copyright 2025 Intel, Ltd.
+ */
+
+#include <preboot.h>
+#include <cmd_module.h>
+
+#define SYS_CLOCK_FREQUENCY	35000000ULL
+
+u64 hl_pll_timestamp_freq_get(void)
+{
+	return SYS_CLOCK_FREQUENCY;
+}
+
+void hl_dev_ts_init(void) {} /* no TS on FPGA */
+
+void hl_ts_shutdown(void) {} /* no TS on FPGA */
+
+enum cmd_check_res dev_specific_module_pci_check(void)
+{
+	/*
+	 * FLR and Hot Reset Registers are not part of PSOC and hence
+	 * writing to them results in ERROR that is resetting the
+	 * ONIX. To avoid it, simply return with NOT FOUND
+	 */
+	return NOT_FOUND;
+}
+
+void module_pci_do_setup(void) {}
+void preboot_plat_early_init(void) {}
+
+u32 hl_skip_pcie_init(void)
+{
+	return 1;
+}
diff --git a/arch/jgs/arc/palad.c b/arch/jgs/arc/palad.c
new file mode 100644
index 0000000..02bbc18
--- /dev/null
+++ b/arch/jgs/arc/palad.c
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier:     Apache-2.0
+/*
+ * (C) Copyright 2025 Intel, Ltd.
+ */
+
+#include <axi_drain_wrap_regs.h>
+#include <axi_drain_wrap_masks.h>
+#include <arc_local_conf_regs.h>
+#include <arc_local_conf_masks.h>
+#include <pcie_dbi_regs.h>
+#include <pcie_dbi_masks.h>
+#include <pcie_aux_regs.h>
+#include <pcie_aux_masks.h>
+#include <pcie_core_regs.h>
+#include <pcie_core_masks.h>
+#include <pcie_flr_regs.h>
+#include <psoc_boot_conf_regs.h>
+#include <psoc_boot_conf_masks.h>
+#include <fs1_blocks_linux_driver.h>
+
+#include <address_space.h>
+#include <preboot.h>
+#include <hl_device.h>
+#include <hl_platform_defs.h>
+#include "hl.h"
+#include <hl_boot_info.h>
+#include <fs1_reg_map.h>
+#include <fs1_efuse.h>
+#include <efuse.h>
+#include <hl_pcie.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(hl_preboot_palad);
+
+/* timestamp fed with constand ref_clk/2 */
+#define TS_CLOCK_FREQUENCY	25000000ULL
+
+#define UART_CLOCK_FREQUENCY	SYS_CLOCK_FREQUENCY
+#define PALAD_UART_BAUDRATE	300
+
+
+u64 hl_pll_timestamp_freq_get(void)
+{
+	return TS_CLOCK_FREQUENCY;
+}
+
+void hl_dev_ts_init(void)
+{
+	/* TODO: implement ts init for JGS */
+}
+
+void hl_ts_shutdown(void)
+{
+	/* TODO: implement ts shutdown for JGS */
+}
+
+#define DRAIN_STAMP_ENA_MASK	(AXI_DRAIN_WRAP_RSP_ERR_DRAIN_STAMP_RD_EN_M | \
+				 AXI_DRAIN_WRAP_RSP_ERR_DRAIN_STAMP_WR_EN_M)
+
+static void hl_psoc_drain_init(void)
+{
+	/* psoc drain is enabled by default, enable drain address stamp logging */
+	CFG_WREG32(mmIO_0_CFG_HALF_0_PSOC_PARC_CFG_MISC_DRAIN_HBW_BASE +
+		   mmAXI_DRAIN_WRAP_RSP_ERR_DRAIN_STAMP, DRAIN_STAMP_ENA_MASK);
+	CFG_WREG32(mmIO_0_CFG_HALF_0_PSOC_PARC_CFG_MISC_DRAIN_LBW_BASE +
+		   mmAXI_DRAIN_WRAP_RSP_ERR_DRAIN_STAMP, DRAIN_STAMP_ENA_MASK);
+}
+
+
+#define EARLY_WRITE_RESP_MASK	(ARC_LOCAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_FORCE_WR_BUF_M |	\
+				 1 << ARC_LOCAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_AWCACHE_OVRD_EN_S |\
+				 1 << ARC_LOCAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_ARCACHE_OVRD_EN_S)
+
+static void hl_lbw_early_write_resp_enable(void)
+{
+	u32 val;
+
+	val = CFG_RREG32(mmIO_0_CFG_HALF_0_PSOC_ARC_MNG_LOCAL_BASE +
+			 mmARC_LOCAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2);
+	val |= EARLY_WRITE_RESP_MASK;
+	CFG_WREG32(mmIO_0_CFG_HALF_0_PSOC_ARC_MNG_LOCAL_BASE +
+		   mmARC_LOCAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2, val);
+}
+
+static bool is_cxl_mode(void)
+{
+	u32 val;
+
+	val = CFG_RREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_H));
+	val &= PSOC_BOOT_CONF_BOOT_STRAP_PINS_H_CXL_MODE_M;
+
+	return !!val;
+}
+
+#define SRIOV_MAX_NUM_VFS	4
+#define SRIOV_DEFAULT_NUM_VFS	1
+
+static void palad_pcie_dynamic_init(void)
+{
+	bool cxl_mode = is_cxl_mode();
+	u32 pci_efuse_cfg;
+	u8 num_vf = 0;
+	u32 val;
+
+	preboot_fuse_read(IO_EFUSE_PALAD_CFG_0_OFFSET, &pci_efuse_cfg);
+	LOG_INF("efuse pci cfg = 0x%x", pci_efuse_cfg);
+
+	if (pci_efuse_cfg & CRS_MODE_ENABLED_M) {
+		/* Set SNNP EP into CRS mode */
+		CFG_RMWREG32(IO0_H0_PCIE_CORE(mmPCIE_CORE_RN), 0, PCIE_CORE_RN_APP_DRS_READY_M);
+		CFG_WREG32(IO0_H0_PCIE_FLR(mmPCIE_FLR_APP_REQ_RETRY_EN), BIT(0));
+		/* TODO: Need to configure APP_PF_REQ_RETRY_EN, APP_VF_REQ_RETRY_EN (SRIOV mode) */
+		LOG_INF("CRS mode enabled in SNPS EP");
+	}
+
+	if (cxl_mode) {
+		LOG_INF("cxl mode");
+		/* CXL capability is enabled as reset default */
+
+		/* disable CXL.Cache capability if required */
+		if (!(pci_efuse_cfg & CXL_CACHE_CAP_ENABLED_M)) {
+			CFG_RMWREG32(
+				IO0_H0_PCIE_DBI(mmPCIE_DBI_CXL_RCRB_DVSEC_HDR_2_FLEXBUS_CAP_OFF),
+				0, PCIE_DBI_CXL_RCRB_DVSEC_HDR_2_FLEXBUS_CAP_OFF_CACHE_CAP_M);
+			CFG_RMWREG32(
+				IO0_H0_PCIE_DBI(mmPCIE_DBI_CXL_RCIEP_DVSEC_HDR_2_FLEXBUS_CAP_OFF),
+				0, PCIE_DBI_CXL_RCIEP_DVSEC_HDR_2_FLEXBUS_CAP_OFF_CACHE_CAP_M);
+		}
+	} else { /* pci mode */
+		LOG_INF("pci mode");
+		/* switch to PCI mode by disabling Alternative Protocol Support */
+		CFG_RMWREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_PL32G_CAPABILITY_REG), 0,
+			     PCIE_DBI_PL32G_CAPABILITY_REG_MOD_TS_ALT_PROTOCOL_SUPPORT_M);
+
+		/* set number of VFs */
+		num_vf = pci_efuse_cfg & SRIOV_NUM_OF_VFS_M;
+		if (num_vf > SRIOV_MAX_NUM_VFS) {
+			num_vf = SRIOV_DEFAULT_NUM_VFS;
+			LOG_WRN("VF exceeds 4, using default of 1 VF");
+		}
+
+		val = num_vf << PCIE_DBI_TOTAL_VFS_INITIAL_VFS_REG_SRIOV_INITIAL_VFS_S |
+		      num_vf << PCIE_DBI_TOTAL_VFS_INITIAL_VFS_REG_SRIOV_TOTAL_VFS_S;
+		CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_TOTAL_VFS_INITIAL_VFS_REG), val);
+		CFG_RMWREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_FLIT_INJECT_CAP_HDR_REG),
+			     mmPCIE_DBI_RESBAR_CAP_HDR_REG,
+			     PCIE_DBI_FLIT_INJECT_CAP_HDR_REG_FNEXT_OFFSET_M);
+	}
+
+	/* skip ATS capability */
+	if (!(pci_efuse_cfg & ATS_CAPABILITY_ENABLED_M)) {
+		val = CFG_RREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_BASE_REG));
+		val &= ~PCIE_DBI_SRIOV_BASE_REG_SRIOV_NEXT_OFFSET_M;
+		val |= mmPCIE_DBI_LTR_CAP_HDR_REG <<
+			PCIE_DBI_SRIOV_BASE_REG_SRIOV_NEXT_OFFSET_S;
+		CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_BASE_REG), val);
+	}
+}
+
+#define BAR_SIZE_1GB		(SZ_1G - 1)
+#define VF_RES_BAR_SIZES	(PCIE_DBI_VF_RESBAR_CAP_REG_2_REG_V512GB_M | \
+				 PCIE_DBI_VF_RESBAR_CAP_REG_2_REG_V256GB_M | \
+				 PCIE_DBI_VF_RESBAR_CAP_REG_2_REG_V128GB_M | \
+				 PCIE_DBI_VF_RESBAR_CAP_REG_2_REG_V64GB_M)
+
+static void palad_pcie_params_init(void)
+{
+	/* enable writing to DBI RO regs */
+	CFG_RMWREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_MISC_CONTROL_1_OFF), 1,
+		     PCIE_DBI_MISC_CONTROL_1_OFF_DBI_RO_WR_EN_M);
+
+	/* switch to shadow registers access in order to set BAR masks */
+	CFG_WREG32(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_DBI, PCIE_AUX_DBI_CS2_M);
+
+	/* set PF CFG BAR to 1GB */
+	CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_BAR0_REG), BAR_SIZE_1GB);
+	CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_BAR1_REG), 0);
+
+	/* set SRIOV CFG BAR to 1GB */
+	CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_BAR0_REG), BAR_SIZE_1GB);
+	CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_BAR1_REG), 0);
+
+	/* disable shadow registers access */
+	CFG_WREG32(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_DBI, 0);
+
+	/* set VF resizable BAR #2 allowed size */
+	CFG_WREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_VF_RESBAR_CAP_REG_2_REG), VF_RES_BAR_SIZES);
+
+	palad_pcie_dynamic_init();
+
+	/* disable writing to DBI RO regs */
+	CFG_RMWREG32(IO0_H0_PCIE_DBI(mmPCIE_DBI_MISC_CONTROL_1_OFF), 0,
+		     PCIE_DBI_MISC_CONTROL_1_OFF_DBI_RO_WR_EN_M);
+}
+
+#define LTTSM_ENA_MASK	(PCIE_AUX_LTSSM_EN_IND_M | PCIE_AUX_LTSSM_EN_MASK_M)
+
+static void hl_palad_pcie_init(void)
+{
+	u32 smlh = CFG_RREG32(pci.aux_smlh_link_up_reg);
+	u32 rdlh = CFG_RREG32(pci.aux_rdlh_link_up_reg);
+
+	/* if link is up, nothing more to do */
+	if (smlh && rdlh)
+		return;
+
+	palad_pcie_params_init();
+
+	CFG_WREG32(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_LTSSM_EN, LTTSM_ENA_MASK);
+
+	/* release request retry so the host can access pci config space */
+	CFG_WREG32(IO0_H0_PCIE_FLR(mmPCIE_FLR_APP_REQ_RETRY_EN), 0);
+	CFG_WREG32(IO0_H0_PCIE_FLR(mmPCIE_FLR_APP_PF_REQ_RETRY_EN), 0);
+	CFG_WREG32(IO0_H0_PCIE_FLR(mmPCIE_FLR_APP_VF_REQ_RETRY_EN), 0);
+}
+
+void preboot_plat_early_init(void)
+{
+	hl_psoc_drain_init();
+	hl_lbw_early_write_resp_enable();
+	hl_palad_pcie_init();
+}
+
+void module_fwupd_dev_spec_shutdown(void) {}
+void module_oob_fwupd_dev_spec_shutdown(void) {}
+
+void preboot_security_init(void)
+{
+}
+
+int get_bmu_development_phase(enum bmu_phase *dev_phase)
+{
+	uint32_t efuse_cfg;
+
+	preboot_fuse_read(IO_EFUSE_PALAD_CFG_0_OFFSET, &efuse_cfg);
+
+	if (efuse_cfg & BMU_INIT_PROD_MODE) {
+		*dev_phase = BMU_PRODUCTION_MODE;
+		LOG_INF("BMU production mode");
+	} else if (efuse_cfg & BMU_INIT_DEVEL_MODE) {
+		*dev_phase = BMU_PRODUCTION_MODE;
+		LOG_INF("BMU development mode");
+	} else {
+		LOG_INF("BMU init disabled, skipping");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+/* In order to improve fit image loading time on palad, the hash calculation is skipped */
+int fit_node_check_hash(const void *fit, int node, const u8 *buf, int len, void *hash)
+{
+	return 0;
+}
diff --git a/arch/jgs/arc/sha.h b/arch/jgs/arc/sha.h
new file mode 100644
index 0000000..42c9893
--- /dev/null
+++ b/arch/jgs/arc/sha.h
@@ -0,0 +1,23 @@
+/* SPDX-License-Identifier:     Apache-2.0
+ *
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#ifndef ARCH_FS1_ARC_SHA_H_
+#define ARCH_FS1_ARC_SHA_H_
+
+#include <sha512.h>
+#include <hash_defs.h>
+#include <preboot/fit_rsa.h>
+
+#define SHA_SUM_LEN			SHA384_SUM_LEN
+#define SHA_DER_LEN			SHA384_DER_LEN
+#define SHA_DER_PRE			sha384_der_prefix
+#define FIT_SHA				"sha384"
+#define FIT_ALGO_PROP_EXPECTED_VALUE	"sha384,rsa3072"
+
+void sha_csum(const unsigned char *input, unsigned int ilen, unsigned char *output);
+
+void sha_csum_regions(const struct image_region region[], int region_count, u8 *output);
+
+#endif /* ARCH_FS1_ARC_SHA_H_ */
diff --git a/arch/jgs/arc/tstd.h b/arch/jgs/arc/tstd.h
new file mode 100644
index 0000000..a3fa1d8
--- /dev/null
+++ b/arch/jgs/arc/tstd.h
@@ -0,0 +1,37 @@
+/* SPDX-License-Identifier:     Apache-2.0
+ *
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#ifndef ARCH_FS1_ARC_TSTD_H_
+#define ARCH_FS1_ARC_TSTD_H_
+
+/* TODO: set proper values fo FS1 */
+#define HL_TS_CMN_SDIF_WDATA_SHIFT	0
+#define HL_TS_CMN_SDIF_WDATA_MASK	0
+#define HL_TS_CMN_SDIF_ADDR_SHIFT	0
+#define HL_TS_CMN_SDIF_ADDR_MASK	0
+#define HL_TS_CMN_SDIF_WRN_SHIFT	0
+#define HL_TS_CMN_SDIF_WRN_MASK		0
+#define HL_TS_CMN_SDIF_PROG_SHIFT	0
+#define HL_TS_CMN_SDIF_PROG_MASK	0
+#define HL_TS_CMN_SDIF_STATUS_BUSY_MASK	0
+
+#define HL_TS_00_SDIF_DATA_SMPL_DATA_MASK 0
+#define HL_TS_CMN_CLK_SYNTH_LO_SHIFT	0
+#define HL_TS_CMN_CLK_SYNTH_LO_MASK	0
+#define HL_TS_CMN_CLK_SYNTH_HI_SHIFT	0
+#define HL_TS_CMN_CLK_SYNTH_HI_MASK	0
+#define HL_TS_CMN_CLK_SYNTH_HOLD_SHIFT	0
+#define HL_TS_CMN_CLK_SYNTH_HOLD_MASK	0
+#define HL_TS_CMN_CLK_SYNTH_ENA_SHIFT	0
+#define HL_TS_CMN_CLK_SYNTH_ENA_MASK	0
+
+#define HL_TS_CMN_SDIF_STATUS		0
+#define HL_TS_CMN_SDIF			0
+#define HL_TS_00_SDIF_DATA		0
+#define HL_TS_CMN_CLK_SYNTH		0
+#define HL_TS_00_ALARMA_CFG		0
+#define HL_TS_00_ALARMB_CFG		0
+
+#endif /* ARCH_FS1_ARC_TSTD_H_ */
diff --git a/arch/jgs/common/common.c b/arch/jgs/common/common.c
new file mode 100644
index 0000000..42c9095
--- /dev/null
+++ b/arch/jgs/common/common.c
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier:     Apache-2.0
+/*
+ * Copyright (C) 2025 Intel Ltd.
+ */
+
+#include <stdio.h>
+
+bool pci_dev_specific_no_ret_check(void)
+{
+	return false;
+}
diff --git a/arch/jgs/common/efuse_parse.c b/arch/jgs/common/efuse_parse.c
new file mode 100644
index 0000000..76c2cf2
--- /dev/null
+++ b/arch/jgs/common/efuse_parse.c
@@ -0,0 +1,403 @@
+// SPDX-License-Identifier:     Apache-2.0
+/*
+ * (C) Copyright 2025 Intel, Ltd.
+ */
+
+#include "common.h"
+#include <preboot.h>
+#include "hl.h"
+#include <zephyr/drivers/misc/common/common_utils.h>
+#include <zephyr/drivers/misc/common/hl_utils.h>
+#include <zephyr/drivers/misc/common/jgs/hl_platform_defs.h>
+#include <zephyr/drivers/misc/common/jgs/jgs_binning_defs.h>
+#include <jgs_fw_if.h>
+#include <jgs_efuse.h>
+#include <gpu_utils.h>
+#include <hl_boot_desc.h>
+#include <hl_boot_info.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/device.h>
+#include <include/zephyr/drivers/efuse.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(efuse_parse);
+
+
+/* TODO: Confirm efuse decoder redundancy value and IO decoder mapping */
+static const s8 vdec_fuse_map[9] = {0, 1, 2, 3, 4, 5, 6, 7, 8};
+
+static int get_redundant_vdec(enum fs1_die_id die_id, u16 fault_mask, s8 *vdec)
+{
+	unsigned int vdec_idx;
+
+	if (hweight16(fault_mask) > 1) {
+		LOG_ERR("multiple faulty decoders in a die, unsupported condition");
+		return -1;
+	}
+
+	if (fault_mask != 0) {
+		vdec_idx = __ffs(fault_mask);
+		*vdec = vdec_fuse_map[vdec_idx];
+	}
+	return 0;
+}
+
+/* TODO: Confirm efuse nic redundancy value and IO nic mapping */
+static const s8 nic_fuse_map[13] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
+
+static int get_redundant_nic(enum fs1_die_id die_id, u16 fault_mask, s8 *val)
+{
+	unsigned int nic_idx;
+
+	if (hweight16(fault_mask) > 1) {
+		LOG_ERR("multiple faulty nic in a die, unsupported condition");
+		return -1;
+	}
+
+	if (fault_mask != 0) {
+		nic_idx = __ffs(fault_mask);
+		*val = nic_fuse_map[nic_idx];
+	}
+	return 0;
+}
+
+static int jgs_vdec_redundancy_ctx_init(struct fs1_redundancy_ctx_io *ctx_io,
+				enum fs1_die_id die_id)
+{
+	int err;
+	uint32_t val;
+
+	err = dev_efuse_get_field(die_id, EFUSE_IO_VDEC_AND_NIC_REDUN_OFFSET,
+				EFUSE_IO_VDEC_REDUN_MASK, &val);
+	if (err)
+		return err;
+
+	/*
+	 * if there is no faulty vdec in efuse use default fault mask, otherwise use
+	 * efuse val
+	 */
+	if (val) {
+		ctx_io->dec_fault_mask[0] = val & IO_VDEC8_FAULT_MASK_D0_L;
+		ctx_io->dec_fault_mask[1] = (val & IO_VDEC8_FAULT_MASK_D0_H) >>
+						IO_VDEC8_NUM_PER_HALF_DIE;
+	}
+
+	/* lower half */
+	err = get_redundant_vdec(die_id, ctx_io->dec_fault_mask[0], &ctx_io->redundant_dec[0]);
+	if (err)
+		return err;
+
+	/* upper half */
+	return get_redundant_vdec(die_id, ctx_io->dec_fault_mask[1], &ctx_io->redundant_dec[1]);
+
+}
+
+static int jgs_nic_redundancy_ctx_init(struct fs1_redundancy_ctx_io *ctx_io, enum fs1_die_id die_id)
+{
+	int err;
+	uint32_t val;
+
+	err = dev_efuse_get_field(die_id, EFUSE_IO_VDEC_AND_NIC_REDUN_OFFSET,
+				EFUSE_IO_NIC_REDUN_MASK, &val);
+	if (err)
+		return err;
+
+	/*
+	 * if there is no faulty vdec in efuse use default fault mask, otherwise use
+	 * efuse val
+	 */
+	if (val)
+		ctx_io->nic_fault_mask = (uint16_t) val;
+
+	return get_redundant_nic(die_id, ctx_io->nic_fault_mask, &ctx_io->redundant_nic);
+}
+
+static int jgs_pmmu_redundancy_ctx_init(struct fs1_redundancy_ctx_io *ctx_io,
+				enum fs1_die_id die_id)
+{
+	bool io_die_stand_alone_mode = dev_is_standalone_io();
+
+	/* PMMU unit is only used in Die 0 in IO-stand-alone package.
+	 * In FS1 product, it is not used, hence binned out
+	 */
+	if ((die_id == FS1_IO_DIE_0) && (io_die_stand_alone_mode)) {
+		ctx_io->redundant_pmmu = FS1_PMMU_IN_USE;  /* will be used */
+		ctx_io->pmmu_fault_mask = 0;
+	} else {
+		/* for IO Die-1 in stand alone and production; and IO Die-0 in production */
+		ctx_io->redundant_pmmu = FS1_PMMU_BINNED_OUT; /* not required; redundant */
+		ctx_io->pmmu_fault_mask = BIT(0);
+	}
+
+	return 0;
+}
+
+static int jgs_redundancy_ctx_io_set_default(void *ctx, enum fs1_die_id die_id)
+{
+	int ret;
+	uint32_t io_die = ((die_id - FS1_IO_DIE_0) / FS1_NUM_HALF_DIES);
+	struct fs1_redundancy_ctx_io *ctx_io = (struct fs1_redundancy_ctx_io *)ctx;
+
+	LOG_INF("%s: io_die = %u", __func__, io_die);
+
+	/* Individual IO die initialization */
+	ctx_io->initialized = true;
+
+	/* always working with 8 vdec per half die and 1 spare per half die */
+	ctx_io->redundant_dec[0] = FS1_FAULTY_VDEC_8;
+	ctx_io->redundant_dec[1] = FS1_FAULTY_VDEC_8;
+	ctx_io->dec_fault_mask[0] = IO_DEC8_FAULT_MASK_DEFAULT;
+	ctx_io->dec_fault_mask[1] = IO_DEC8_FAULT_MASK_DEFAULT;
+
+	/* always working with 12 NIC per die and 1 spare */
+	ctx_io->redundant_nic = FS1_FAULTY_NIC_12;
+	ctx_io->nic_fault_mask = IO_NIC12_FAULT_MASK_DEFAULT;
+
+	ret = jgs_vdec_redundancy_ctx_init(ctx_io, die_id);
+	if (ret) {
+		LOG_ERR("vdec redundancy ctx init failed");
+		return ret;
+	}
+
+	ret = jgs_nic_redundancy_ctx_init(ctx_io, die_id);
+	if (ret) {
+		LOG_ERR("nic redundancy ctx init failed");
+		return ret;
+	}
+
+	ret = jgs_pmmu_redundancy_ctx_init(ctx_io, die_id);
+	if (ret) {
+		LOG_ERR("pmmu redundancy ctx init failed");
+		return ret;
+	}
+
+	/* Note: pmmu isolation and binning does not depend on efuse values.
+	 * These are dependent on use case/design based, hence binning and isolation
+	 * will be done accordingly.
+	 */
+
+	/* TODO: update API with  bootinfo/littlefs */
+	ret = hl_boot_info_set_partial(HL_BOOT_INFO_BINNING_AND_ISO, ctx_io,
+					sizeof(struct fs1_redundancy_ctx_io) * io_die,
+					sizeof(struct fs1_redundancy_ctx_io));
+	if (ret)
+		LOG_ERR("%s: Failed to store redundancy ctx to bootinfo ret=%d", __func__, ret);
+
+	return ret;
+}
+
+/* TODO: Confirm efuse xecore redundancy value and IO xecore mapping */
+static const s8 xecore_fuse_map[17] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+
+static int get_redundant_xecore(enum fs1_die_id die_id, u32 fault_mask, s8 *val)
+{
+	unsigned int xecore_idx;
+
+	if (hweight16(fault_mask) > 1) {
+		LOG_ERR("multiple faulty xecore in a die, unsupported condition");
+		return -1;
+	}
+
+	if (fault_mask != 0) {
+		xecore_idx = __ffs(fault_mask);
+		*val = xecore_fuse_map[xecore_idx];
+	}
+	return 0;
+}
+
+static int jgs_xecore_redundancy_ctx_init(struct fs1_redundancy_ctx_compute *ctx_compute,
+				enum fs1_die_id die_id)
+{
+	int err;
+	uint32_t val;
+
+	/* TODO: PRM is not updated and 3.4 (17 XECORE) and 7.2.6 (15 XECORE) contradictory */
+	/* Assuming 16 XECORE */
+	uint32_t efuse_xecore_redn_offset = (die_id <= FS1_COMPUTE_DIE_1) ?
+						EFUSE_COMPUTE_XECORE_0_REDUN_OFFSET :
+						EFUSE_COMPUTE_XECORE_1_REDUN_OFFSET;
+	 /* TODO: correct mask */
+	uint32_t efuse_xecore_redn_mask = (die_id % 2) ? 0x0000FFFF : 0xFFFF0000;
+
+	err = dev_efuse_get_field(die_id, efuse_xecore_redn_offset, efuse_xecore_redn_mask, &val);
+	if (err)
+		return err;
+
+	/*
+	 * if there is no faulty xecore in efuse use default fault mask, otherwise use
+	 * efuse val
+	 */
+	if (val)
+		ctx_compute->xecore_fault_mask = val;
+
+	return get_redundant_xecore(die_id, ctx_compute->xecore_fault_mask,
+					&ctx_compute->redundant_xecore);
+}
+
+/* TODO: Confirm efuse pdma redundancy value and IO pdma mapping */
+static const s8 pdma_fuse_map[2] = {0, 1};
+
+static int get_redundant_pdma(enum fs1_die_id die_id, uint8_t fault_mask, s8 *val)
+{
+	unsigned int pdma_idx;
+
+	if (hweight8(fault_mask) > 1) {
+		LOG_ERR("multiple faulty pdma in a die, unsupported condition");
+		return -1;
+	}
+
+	if (fault_mask != 0) {
+		pdma_idx = __ffs(fault_mask);
+		*val = pdma_fuse_map[pdma_idx];
+	}
+	return 0;
+}
+
+static int jgs_pdma_redundancy_ctx_init(struct fs1_redundancy_ctx_compute *ctx_compute,
+				enum fs1_die_id die_id)
+{
+	int err;
+	uint32_t val;
+
+	err = dev_efuse_get_field(die_id, EFUSE_COMPUTE_PDMA_REDN_OFFSET,
+					EFUSE_COMPUTE_PDMA_REDN_MASK, &val);
+	if (err)
+		return err;
+
+	/*
+	 * if there is no faulty pdma in efuse use default fault mask, otherwise use
+	 * efuse val
+	 */
+	if (val)
+		ctx_compute->pdma_fault_mask = (uint8_t) val;
+
+	return get_redundant_pdma(die_id, ctx_compute->pdma_fault_mask,
+				&ctx_compute->redundant_pdma);
+}
+
+/* TODO: Confirm efuse edma redundancy value and IO edma mapping */
+static const s8 edma_fuse_map[5] = {0, 1, 2, 3, 4};
+
+static int get_redundant_edma(enum fs1_die_id die_id, uint8_t fault_mask, s8 *val)
+{
+	unsigned int edma_idx;
+
+	if (hweight8(fault_mask) > 2) {
+		LOG_ERR("more than 2 faulty edma in a die, unsupported condition");
+		return -1;
+	}
+
+	if (fault_mask != 0) {
+		/* 1st redundant */
+		edma_idx = __ffs(fault_mask);
+		val[0] = edma_fuse_map[edma_idx];
+
+		/* 2nd redundant */
+		fault_mask &= ~(1 << edma_idx); /* clear MSB, check for next faulty edma */
+		if (fault_mask == 0) /* for single faulty edma in efuse, generate 2nd redundant */
+			/* clear already redundant one done above */
+			fault_mask = EFUSE_COMPUTE_EDMA_REDN_MASK & ~(1 << edma_idx);
+
+		edma_idx = __ffs(fault_mask);
+		val[1] = edma_fuse_map[edma_idx];
+	}
+
+	return 0;
+}
+
+static int jgs_edma_redundancy_ctx_init(struct fs1_redundancy_ctx_compute *ctx_compute,
+				enum fs1_die_id die_id)
+{
+	int err;
+	uint32_t val;
+
+	err = dev_efuse_get_field(die_id, EFUSE_COMPUTE_EDMA_REDN_OFFSET,
+				EFUSE_COMPUTE_EDMA_REDN_MASK, &val);
+	if (err)
+		return err;
+
+	/*
+	 * if there is no faulty edma in efuse use default fault mask, otherwise use
+	 * efuse val
+	 */
+	if (val)
+		ctx_compute->edma_fault_mask = (uint8_t)val;
+
+	return get_redundant_edma(die_id, ctx_compute->edma_fault_mask,
+				&ctx_compute->redundant_edma[0]);
+}
+
+static int jgs_redundancy_ctx_compute_set_default(void *ctx, enum fs1_die_id compute_die)
+{
+	int ret;
+	struct fs1_redundancy_ctx_compute *ctx_compute = (struct fs1_redundancy_ctx_compute *)ctx;
+
+	ctx_compute->initialized = true;
+
+	/* always working with 3 EDMA per die and 2 spare */
+	ctx_compute->redundant_edma[0] = FS1_FAULTY_EDMA_4;
+	ctx_compute->redundant_edma[1] = FS1_FAULTY_EDMA_3;
+	ctx_compute->edma_fault_mask = COMPUTE_EDMA_3_4_FAULT_MASK_DEFAULT;
+
+	/* always working with 2 PDMA per die and 1 spare */
+	ctx_compute->redundant_pdma = FS1_FAULTY_PDMA_2;
+	ctx_compute->pdma_fault_mask = COMPUTE_PDMA_2_FAULT_MASK_DEFAULT;
+
+	/* always working with 16 xecore per die and 1 spare */
+	ctx_compute->redundant_xecore = FS1_FAULTY_XECORE_16;
+	ctx_compute->xecore_fault_mask = COMPUTE_XECORE_17_FAULT_MASK_DEFAULT;
+
+	/* TODO: 1 CS + 6x HBM PC , AKA Dual plane. */
+
+	ret = jgs_edma_redundancy_ctx_init(ctx_compute, compute_die);
+	if (ret) {
+		LOG_ERR("edma redundancy ctx init failed");
+		return ret;
+	}
+
+	ret = jgs_pdma_redundancy_ctx_init(ctx_compute, compute_die);
+	if (ret) {
+		LOG_ERR("pdma redundancy ctx init failed");
+		return ret;
+	}
+
+	ret = jgs_xecore_redundancy_ctx_init(ctx_compute, compute_die);
+	if (ret) {
+		LOG_ERR("xecore redundancy ctx init failed");
+		return ret;
+	}
+
+	/* TODO: 1 CS + 6x HBM PC , AKA Dual plane. */
+
+	/* TODO: update API with  bootinfo/littlefs */
+	ret = hl_boot_info_set_partial(HL_BOOT_INFO_BINNING_AND_ISO, ctx_compute,
+					(sizeof(struct fs1_redundancy_ctx_io) * FS1_NUM_IO_DIES +
+					sizeof(struct fs1_redundancy_ctx_compute) * compute_die),
+					sizeof(struct fs1_redundancy_ctx_compute));
+
+	if (ret)
+		LOG_ERR("%s: Failed to store redundancy ctx to bootinfo ret=%d", __func__, ret);
+
+	return ret;
+}
+
+int jgs_redundancy_ctx_init(void *ctx, enum fs1_die_id die_id)
+{
+	int ret;
+
+	LOG_INF("%s: die_id = %u", __func__, die_id);
+
+	if (die_id <= FS1_COMPUTE_DIE_3)
+		ret = jgs_redundancy_ctx_compute_set_default(ctx, die_id);
+	else if (die_id <= FS1_IO_DIE_1_HALF_1)
+		ret = jgs_redundancy_ctx_io_set_default(ctx, die_id);
+	else {
+		LOG_ERR("%s: invalid die id = %u", __func__, die_id);
+		return -1;
+	}
+
+	return ret;
+}
+
diff --git a/arch/jgs/prj.conf b/arch/jgs/prj.conf
new file mode 100644
index 0000000..ba1d71e
--- /dev/null
+++ b/arch/jgs/prj.conf
@@ -0,0 +1,7 @@
+CONFIG_PREBOOT_MODULE_PCI=y
+CONFIG_PREBOOT_MODULE_LKD=y
+CONFIG_PREBOOT_MODULE_FWUPD=y
+CONFIG_PREBOOT_MODULE_FWUPD_OOB=n
+CONFIG_PREBOOT_MODULE_I2C_S=n
+CONFIG_NUM_METAIRQ_PRIORITIES=1
+CONFIG_USE_WATCHDOG_TIMER_API=y
diff --git a/lib/rma.c b/lib/rma.c
index 0a4c2b7..c0d7a2c 100644
--- a/lib/rma.c
+++ b/lib/rma.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier:     Apache-2.0
 /*
- * (C) Copyright 2022 HabanaLabs, Ltd.
+ * (C) Copyright 2025 Intel, Ltd.
  */
 
 
@@ -9,6 +9,9 @@
 #ifdef CONFIG_SOC_FS1_PREBOOT
 #include <cpucp_common_if.h>
 #include <hl_boot_if.h>
+#elif CONFIG_SOC_JGS_PREBOOT
+#include <cpucp_common_if.h>
+#include <hl_boot_if.h>
 #else
 #include <cpucp_if.h>
 #endif
